zipkin/src/main/java/com/linecorp/armeria/common/tracing/RequestContextCurrentTraceContext.java

```diff
@@ -0,0 +1,157 @@
+     * }</pre>
+     */
+    public static void copy(RequestContext src, RequestContext dst) {
+        dst.attr(TRACE_CONTEXT_KEY).set(src.attr(TRACE_CONTEXT_KEY).get());
+    }
+
+    @Override
+    @Nullable
+    public TraceContext get() {
+        final Attribute<TraceContext> traceContextAttribute = getTraceContextAttributeOrWarnOnce();
+        return traceContextAttribute != null ? traceContextAttribute.get() : null;
+    }
+
+    @Override
+    public Scope newScope(@Nullable TraceContext currentSpan) {
+        // Handle inspection added to ensure we can fail-fast if this isn't installed.
+        if (currentSpan != null && PingPongExtra.maybeSetPong(currentSpan)) {
+            return Scope.NOOP;
+        }
+
+        final Attribute<TraceContext> traceContextAttribute = getTraceContextAttributeOrWarnOnce();
+        if (traceContextAttribute == null) {
+            return INCOMPLETE_CONFIGURATION_SCOPE;
+        }
+
+        final TraceContext previous = traceContextAttribute.getAndSet(currentSpan);
+
+        // Don't remove the outer-most context (client or server request)
+        if (previous == null) {
+            return INITIAL_REQUEST_SCOPE;
+        }
+
+        // Removes sub-spans (i.e. local spans) from the current context when Brave's scope does.
+        // If an asynchronous sub-span, it may still complete later.
+        class RequestContextTraceContextScope implements Scope {
+            @Override
+            public void close() {
+                // re-lookup the attribute to avoid holding a reference to the request if this scope is leaked
+                getTraceContextAttributeOrWarnOnce().set(previous);
+            }
+        }
+
+        return new RequestContextTraceContextScope();
+    }
+
+    /** Armeria code should always have a request context available, and this won't work without it. */
+    private static Attribute<TraceContext> getTraceContextAttributeOrWarnOnce() {
+        return RequestContext.mapCurrent(r -> r.attr(TRACE_CONTEXT_KEY), LogRequestContextWarningOnce.INSTANCE);
+    }
+
+    private RequestContextCurrentTraceContext() {
+    }
+
+    private enum LogRequestContextWarningOnce implements Supplier<Attribute<TraceContext>> {
+
+        INSTANCE;
+
+        @Override
+        public Attribute<TraceContext> get() {
+            ClassLoaderHack.loadMe();
+            return null;
+        }
+
+        /**
+         * This won't be referenced until {@link #get()} is called. If there's only one classloader, the
+         * initializer will only be called once.
+         */
+        static class ClassLoaderHack {
+            static void loadMe() {
+            }
+
+            static {
+                logger.warn("Attempted to propagate trace context, but no request context available. " +
+                            "Did you remember to use RequestContext.contextAwareExecutor() or " +
+                            "RequestContext.makeContextAware()");
```

Comment:
- Missing trailing question mark? i.e. `RequestContext.makeContextAware()?`
- 'Did you forget to...' sounds better to me. What do you think?

Reply:
sgtm